### PR TITLE
Speed up YOLO iOS per-frame postprocessing and run inference on the Apple Neural Engine

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Profiling showed the YOLO26 CoreML models are already near-optimal on the Apple Neural Engine (ANE), so the wins here are in **Swift-side per-frame postprocessing** (array/object marshalling around the decode) plus **compute-unit selection**. Each change was validated with a `swiftc -O` micro-benchmark and locked with a test.

## Changes

| Area | Change | Before → After |
|---|---|---|
| **segment** | Per-instance probability maps (`Masks.masks`, `[[[Float]]]`) built via element-wise nested-array subscripting → bulk row copies from the contiguous buffer | **2.35 ms → 0.18 ms** (30×160²) |
| **segment** | Per-detection mask coefficients in a heavyweight `MLMultiArray` → plain `[Float]` (also drops a force-`try`) | **0.144 ms → 0.026 ms** (300 dets) |
| **semantic** | Per-pixel argmax over NCHW logits (each class read `H*W` apart, cache-thrashing) → cache-friendly class-major pass | **1.11 ms → 0.80 ms** (19×320²) |
| **classify** | Top-5 by full vector sort (`O(n log n)` + tuple-array alloc) → linear pass with a 5-element insertion buffer | **0.057 ms → 0.001 ms** (1000 cls) |
| **all tasks** | `MLModelConfiguration.computeUnits` `.all` → `.cpuAndNeuralEngine` | see below |

The decode-path changes are bit-identical (verified in-benchmark and by tests); public API is unchanged.

### Compute units (`.all` → `.cpuAndNeuralEngine`)
`.all` lets CoreML also schedule work on the GPU. In the live camera app the GPU is already busy compositing the preview and overlays, so GPU scheduling risks contention and frame-time jitter. `.cpuAndNeuralEngine` keeps the conv backbone on the Neural Engine — the engine these models already prefer — so results are unchanged in practice. **Confirmed working on a physical iPhone.**

## Also fixed
- `generateCombinedMaskImage` crashed on **zero detections** (force-unwrapped an empty buffer's base address). Now guards `N == 0` and binds the base address safely, returning no image and empty masks. Covered by a new test.

## Ruled out (no change)
- **detect / pose / OBB postprocessing**: end2end decode paths are already sub-microsecond (raw `dataPointer` + strides, ≤300 dets).
- **export (ultralytics)**: palettization = size-only; int8 activations = no ANE gain; removing the in-graph topk/gather = ~4%; newer `minimum_deployment_target` *regresses* latency. The exported model is already ANE-optimal.

## Testing
`xcodebuild ... clean build test` on the iOS Simulator: **74 tests, 0 failures**. Camera app verified on a physical iPhone. New tests: `PlotTests` (mask output + empty detections), `SemanticSegmenterTests` (argmax), `ClassifierTests` (top-5 vs full-sort).

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR improves real-time iOS YOLO inference performance and stability 📱⚡ by optimizing CoreML compute settings, speeding up classification and segmentation post-processing, and adding tests for correctness ✅

### 📊 Key Changes
- Switched CoreML inference from `.all` to `.cpuAndNeuralEngine` in `BasePredictor` 🧠
  - This avoids GPU contention in live camera apps, where the GPU is already busy rendering the preview and overlays.
- Optimized classifier top-5 prediction logic in `Classifier` 🚀
  - Replaced full sorting of all class scores with a faster single-pass top-5 selection.
  - Made `softmaxProbs()` accessible to tests for validation.
- Reduced segmentation overhead in `Segmenter` ✂️
  - Replaced per-detection `MLMultiArray` mask coefficient storage with lightweight `[Float]` arrays.
  - Updated related parsing, box building, and NMS data flow to use simpler arrays.
- Improved mask generation in `Plot` 🎨
  - Added an early return for empty detections to prevent unsafe buffer access and potential crashes.
  - Switched per-instance mask extraction to bulk row copies, which is much faster than element-by-element copying.
- Accelerated semantic segmentation post-processing in `SemanticSegmenter` 🖼️
  - Reworked class selection into a more cache-friendly two-phase process:
    - first compute the best class per pixel
    - then render colors in one sequential pass
  - Removed the older per-pixel helper method.
- Added new tests for reliability 🧪
  - New classifier test checks that optimized top-5 results match the original full-sort behavior.
  - New semantic segmentation test validates correct per-pixel class selection.
  - Expanded plot tests to verify mask extraction output and safe handling of empty detections.

### 🎯 Purpose & Impact
- Better real-time camera performance on iPhone 📱
  - Prioritizing the Apple Neural Engine helps reduce frame drops and jitter during live inference.
- Lower CPU and memory overhead ⚡
  - Using plain Swift arrays instead of `MLMultiArray` for mask coefficients cuts object creation costs.
- Faster post-processing for classification and segmentation ⏱️
  - These changes should improve responsiveness, especially for repeated frame-by-frame predictions.
- More robust app behavior 🛡️
  - Empty-detection handling reduces crash risk in edge cases.
- Greater confidence in correctness ✅
  - Added tests help ensure the new optimizations preserve output accuracy while improving speed.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolo-ios-app` 🚀 PR! This is an automated message to help streamline review; an engineer will assist shortly. Please review the checklist below to help ensure smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolo-ios-app/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolo-ios-app` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: Sources/YOLO/Plot.swift [hunk 1]
```diff
@@ -321,16 +312,27 @@ func generateCombinedMaskImage(
...
-      for y in 0..<outputHeight {
-        for x in 0..<outputWidth {
-          masksArray[i][y][x] = combinedMask[startIdx + (outputY + y) * maskWidth + outputX + x]
+  // 10) Per-instance probability maps. Copy each output row in one bulk operation from the contiguous
+  //     `combinedMask` buffer instead of writing element-by-element through `[[[Float]]]` subscripts. The
+  //     nested-array form forces per-element ARC/COW/bounds-check overhead; bulk row copies are ~13x faster
+  //     for typical sizes (e.g. 30 instances × 160×160) while producing bit-identical values.
+  if returnIndividualMasks {
+    probabilityMasks = combinedMask.withUnsafeBufferPointer { buf -> [[[Float]]] in
+      let base = buf.baseAddress!
```
Comment (UltralyticsAssistant) [inline Sources/YOLO/Plot.swift]: ⚠️ **HIGH**: This force-unwrap regresses the empty-detections path. When `N == 0`, `combinedMask` is empty, `withUnsafeBufferPointer` provides a `nil` base address, and `returnIndividualMasks == true` will now crash instead of returning an empty mask list like the previous implementation did.

**Suggested change:**
```suggestion
      guard let base = buf.baseAddress else { return [] }
```

File: scripts/perf/coreml_profile.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+for _ in range(5):
+    m.predict(feed)
+ts = []
+for _ in range({runs}):
+    t = time.perf_counter(); m.predict(feed); ts.append((time.perf_counter() - t) * 1000)
+ts.sort()
+print(json.dumps({{"unit": {unit!r}, "runs": {runs}, "best": ts[0], "median": ts[len(ts)//2],
+                   "p90": ts[int(len(ts)*0.9)], "mean": sum(ts)/len(ts)}}))
+"""
+    proc = subprocess.run([sys.executable, "-c", code], capture_output=True, text=True)
```
Comment (UltralyticsAssistant) [inline scripts/perf/coreml_profile.py]: 💡 **MEDIUM**: `_time_one_unit()` treats every child-process failure as `None`, and `main()` later reports that as the known host compiler crash. That masks real errors like an invalid model path, missing dependencies, or unsupported input types, which will make this tool very hard to diagnose when it fails for ordinary reasons. Please surface non-zero exit status / stderr separately and only classify the specific compiler-crash case as ignorable.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The mask-copy optimization and regression test look good overall, but there is one real correctness regression in the Swift fast path for empty detections, and the new profiler currently hides ordinary failures behind a misleading crash message.

💬 Posted 2 inline comments

File: Sources/YOLO/Plot.swift (hunk 1 above)
Comment (glenn-jocher) [reply Sources/YOLO/Plot.swift]: Fixed. Added an early `guard N > 0 else { return (nil, returnIndividualMasks ? [] : nil) }` so the empty-detection case never reaches the unsafe-buffer paths (this also covers the pre-existing `vDSP_mmul` force-unwraps), and changed the row-copy to `guard let base = buf.baseAddress else { return [] }` per your suggestion. Locked with `testGenerateCombinedMaskImageEmptyDetectionsReturnsGracefully`.

File: scripts/perf/coreml_profile.py (hunk 2 above)
Comment (glenn-jocher) [reply scripts/perf/coreml_profile.py]: Good catch. I've removed `scripts/perf/` from this PR to keep it focused on the inference changes, so this no longer applies here. If the profiler is reintroduced as its own PR, the error handling will distinguish the specific MPSGraph host-compiler crash from ordinary failures (bad path / missing deps / unsupported input) and surface stderr.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Huge thanks to @glenn-jocher for pushing this forward.

> “Efficiency is doing better what is already being done.” — Peter Drucker

This update is a great example of that idea in action: smarter CoreML compute selection, faster classification and segmentation post-processing, safer empty-detection handling, and stronger test coverage all add up to a more responsive and reliable real-time iOS YOLO experience. Really appreciate the thoughtful performance work and the attention to correctness behind it. 🚀📱✅
